ARROW-5510: [C++][Python][R][GLib] Implement Feather "V2" using Arrow IPC file format

wesm opened the pull request:
This is based on top of ARROW-7979, so I will need to rebase once that is merged.

Excluding the changes from ARROW-7979, this patch is a substantial code reduction in Feather-related code. I removed a lot of cruft from the V1 implementation and made things a lot simpler without altering the user-facing functionality.

To summarize:

* V2 is exactly the Arrow IPC file format, with the option for the experimental "trivial" body buffer compression implemented in ARROW-7979. `read_feather` functions distinguish the files based on the magic bytes at the beginning of the file ("FEA1" versus "ARROW1")
* A `ipc::feather::WriteProperties` struct has been introduced to allow setting the file version, as well as chunksize (since large tables are broken up into smaller chunks when writing), compression type, and compression level (compressor-specific)
* LZ4 and ZSTD are the only codecs intended to be supported (also in line with mailing list discussion about IPC compression). The default is LZ4 unless -DARROW_WITH_LZ4=OFF in which case it's uncompressed
* Unit tests in Python now test both versions
* R tests are only running the V2 version. I'll need some help adding options to set the version as well as the compression type and compression level

Since 0.17.0 is likely to be released without formalizing IPC compression, I will plan to support an "ARROW:experimental_compression" metadata member in 0.17.0 Feather files. 

Other notes:

* Column decompression is currently serial. I'll work on making this parallel ASAP as it will impact benchmarks significantly. 
* Compression (both chunk-level and column-level) is serial. Write performance would be much improved, especially at higher compression levels, by compressing in parallel at least at the column level
* Write performance could be improved by compressing chunks and writing them to disk concurrently. It's done serially at the moment, so will open a follow up JIRA about this

wesm commented:
I'll work on some rough compression benchmarks with LZ4 and ZSTD using the datasets in https://ursalabs.org/blog/2019-10-columnar-perf/ to see how things are looking as far as file size and load times.

nealrichardson commented:
The R build would be fixed by running `cd r && make doc` (or, not to beat a dead horse, by merging #6411)

wesm commented:
OK, I ran some simple benchmarks on my laptop (8-core i9 processor -- note that BOTH compression and decompression are SINGLE-THREADED currently) with the Fannie Mae and NYC Taxi datasets. The cases are:

* Uncompressed IPC file
* ZSTD with level 1 and 10
* LZ4 (no configurable compression level)

See file sizes:

```
-rw------- 1 wesm wesm  638796850 Mar 23 17:49 2016Q4_lz4_None.feather
-rw-r--r-- 1 wesm wesm 1942965715 Dec  9 21:04 2016Q4.txt
-rw------- 1 wesm wesm 5045771154 Mar 23 17:49 2016Q4_uncompressed_None.feather
-rw------- 1 wesm wesm  395365770 Mar 23 17:49 2016Q4_zstd_10.feather
-rw------- 1 wesm wesm  524043698 Mar 23 17:49 2016Q4_zstd_1.feather
-rw-r--r-- 1 wesm wesm 2728058790 Mar 23 16:46 yellow_tripdata_2010-01.csv
-rw------- 1 wesm wesm 1175453106 Mar 23 17:51 yellow_tripdata_2010-01_lz4_None.feather
-rw------- 1 wesm wesm 2505808570 Mar 23 17:50 yellow_tripdata_2010-01_uncompressed_None.feather
-rw------- 1 wesm wesm  651796626 Mar 23 17:51 yellow_tripdata_2010-01_zstd_10.feather
-rw------- 1 wesm wesm  821963122 Mar 23 17:50 yellow_tripdata_2010-01_zstd_1.feather
```

* ZSTD achieves 90+% compression ratio on Fannie Mae and 60-75% on NYC Taxi 
* LZ4 has lower compression ratio, 87% on Fannie Mae and 54% on NYC Taxi

Here are the load times (note again that decompression is single-threaded):

```
# Load to Arrow Table

('fanniemae', None, None, 0.016037464141845703)  # MEMORY MAP
('fanniemae', 'zstd', 1, 2.873652696609497)
('fanniemae', 'zstd', 10, 1.4272994995117188)
('fanniemae', 'lz4', None, 0.8573403358459473)
('nyctaxi', None, None, 0.004047393798828125)  # MEMORY MAP
('nyctaxi', 'zstd', 1, 2.758413553237915)
('nyctaxi', 'zstd', 10, 2.0100138187408447)
('nyctaxi', 'lz4', None, 0.8240318298339844)

# Load to Arrow Table, convert to pandas

('fanniemae', None, None, 2.4117162227630615)
('fanniemae', 'zstd', 1, 5.116245985031128)
('fanniemae', 'zstd', 10, 3.9139928817749023)
('fanniemae', 'lz4', None, 3.5294902324676514)
('nyctaxi', None, None, 7.1993725299835205)
('nyctaxi', 'zstd', 1, 10.147839069366455)
('nyctaxi', 'zstd', 10, 8.913217782974243)
('nyctaxi', 'lz4', None, 8.480979204177856)
```

This looks pretty excellent to me, a huge benefit to users with less than 2x slowdown when considering conversion to `pandas.DataFrame`.

wesm commented:
This is rebased and should be easier to review now

jorisvandenbossche commented on python/pyarrow/feather.py:
```diff
@@ -175,15 +123,38 @@ def write_feather(df, dest):
         Dataframe to write out as feather format.
```
update there that it also accepts a table? 

(actually, in the parquet module, it uses a different naming scheme, i.e. `write_table`)

jorisvandenbossche commented on python/pyarrow/feather.py:
```diff
@@ -29,74 +29,6 @@ def _check_pandas_version():
         raise ImportError("feather requires pandas >= 0.17.0")
 
 
-class FeatherReader(ext.FeatherReader):
```
Any idea if there would be people using those classes?

(I have no problem with removing them directly, though, but we could also deprecate first)

wesm replied:
I don't think anyone should be using them. Some of the features that they used have been removed

jorisvandenbossche commented on python/pyarrow/feather.py:
```diff
@@ -203,15 +174,37 @@ def read_feather(source, columns=None, use_threads=True):
         Only read a specific set of columns. If not provided, all columns are
         read.
+    as_pandas : bool, default True
+        Convert result to pandas.DataFrame
```
Personally, I would not add this, as this seems inconsistent with the rest of pyarrow (or do we have this keyword elsewhere?), and the alternative is just a `.to_pandas()` call afterwards (so not even less typing). 
(and also, for pandas users, they should probably rather use `pandas.read_feather`)

jorisvandenbossche replied:
Whoops, sorry, I missed that this method actually already returns a pandas DataFrame :) (default of True).

In that case, I would propose to put the bulk of the function in `read_table` (since in principle you should be able to read a feather file into a arrow table without having pandas installed), and have `read_feather` as a small function calling `read_table(...).to_pandas(use_threads=use_threads)`

wesm replied:
Makes sense, will do

wesm commented:
I'll work on fixing the CI issues today. @nealrichardson if you could help me with exposing the version and compression options in R that is the last thing that's needed for R I think

wesm commented on python/pyarrow/feather.py:
```diff
@@ -165,25 +97,68 @@ def read_pandas(self, columns=None, use_threads=True):
+        the default, which is currently 64K
+    version : int, default 2
+        Feather file version. Version 2 is the current. Version 1 is the more
+        limited legacy format
```
@nealrichardson these are the new options that need to be exposed in R. It shouldn't be necessary to wrap any new C++ classes, only adding options to `write_feather`

nealrichardson commented on cpp/src/arrow/ipc/feather.h:
```diff
@@ -116,55 +93,42 @@ class ARROW_EXPORT TableReader {
-  class ARROW_NO_EXPORT TableWriterImpl;
-  std::unique_ptr<TableWriterImpl> impl_;
+  /// Compression type to use
+  Compression::type compression = Compression::ZSTD;
```
Is this correct: zstd compressed by default? Since `ARROW_IPC` does not require `ARROW_WITH_ZSTD`, the default may very well fail (as it turns out from what I'm seeing locally).

Note that the python bindings (docstring at least) say uncompressed is default.

wesm replied:
Ah I'll fix that. If anything LZ4 should probably be the default since based on benchmarks it has little effect on read/write times.

I'll change the implementation to set LZ4 as the default if it's available otherwise uncompressed is the default.

wesm replied:
Done

wesm commented:
Some refactoring is required in GLib (I forgot that GLib had Feather bindings). I'll try to do it myself.

nealrichardson commented on python/pyarrow/feather.py:
```diff
@@ -165,25 +97,68 @@ def read_pandas(self, columns=None, use_threads=True):
+        Use a compression level particular to the chosen compressor. If None
+        use the default compression level
+    chunksize : int, default None
+        For V2 files, the size of chunks to split the data into. None means use
```
```suggestion
        For V2 files, the number of rows each chunk in the file should have. None means use
```

wesm commented:
@kou I just pushed changes to the GLib and Ruby bindings that follow this PR. I removed the FeatherTableWriter class and a number of reader methods that I removed in this patch without deprecation. My reasoning was that users are primarily interacting with these files as a one-shot operation, where they either read the whole file or a subset of columns using names or indices. If you would like to expose the new `ipc::feather::WriteProperties` I will leave that to you for follow up work. 

Feel free to make any changes you feel are appropriate. I did the bare minimum to get the test suite passing.

nealrichardson commented on r/R/feather.R:
```diff
@@ -43,60 +87,15 @@ write_feather <- function(x, sink) {
-
-  writer <- FeatherTableWriter$create(sink)
-  ipc___TableWriter__RecordBatch__WriteFeather(writer, x)
+  ipc___WriteFeather__RecordBatch(sink, x, version, chunk_size, compression, compression_level)
```
@wesm It's odd to me that the writer seems only to write RecordBatches, but the reader returns a Table. Is that just a shortcoming of the bindings?

wesm replied:
It's a shortcoming yeah. It would be good to change it to work with either RecordBatch or Table

wesm replied:
I'm changing things to work with either Table or RecordBatch.

nealrichardson commented:
What happens if I try to write to Feather V1 data types that aren't supported? Does it tell me that I should use V2?

wesm commented:
@nealrichardson it'll fail with an unsupported type error. I can amend the error message to say "use the V2 format, Luke!"

nealrichardson commented:
FYI 

```
> test_check("arrow")
── 1. Error: feather read/write round trip (@test-feather.R#74)  ───────────────
Invalid: LZ4 doesn't support setting a compression level.
Backtrace:
 1. arrow:::expect_feather_roundtrip(...)
 2. arrow:::write_fun(tib, tf2)
 3. arrow::write_feather(x, f, compression = "lz4", compression_level = 3)
 4. arrow:::ipc___WriteFeather__RecordBatch(...)
```

Is this expected? So compression_level should only be a valid option if using zstd?

wesm commented:
> Is this expected? So compression_level should only be a valid option if using zstd?

Right. I will make it so that the LZ4 compression level is ignored anyhow, that is a nuisance.

kou commented on c_glib/test/test-feather-file-reader.rb:
```diff
@@ -18,26 +18,11 @@
-      ensure
-        writer.close
-      end
+      Arrow::feather_write_file(table, output)
```
This API is strange in GLib.
I'll improve this API.

wesm replied:
Thank you, I wasn't sure how to handle non-classes in GLib, didn't see too many examples

wesm replied:
Sorry about the Ruby failure, I am not set up locally to run the red-arrow tests but I will do that soon so I can handle these issues better in the future

kou replied:
No problem. I'll also fix the Ruby failure.

kou replied:
Done.
I'll improve `GArrowFeatherWriteProperties` in a follow-up task.

wesm commented:
Thanks everyone for the assistance. I'll make a couple of refinements per comments above and then merge this

wesm commented:
@nealrichardson while I'm waiting for the CI to run, if you have a moment to take a look at my R changes:

* Accept either data.frame, arrow::Table or arrow::RecordBatch in write_feather
* Use LZ4 as default compression if it's available (this is the Python and C++ library default also)

nealrichardson commented:
@wesm LGTM. Only note was that you could probably collapse the data.frame/record batch to table conversion to a single check, but then you just did it :)

wesm commented:
+1. Thanks all
